### PR TITLE
IGNITE-5217: Add Gradient Descent and QR-based trainers for Linear Regression

### PR DESCRIPTION
**Pull request by dmitrievanthony**

### CODE REVIEW

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+
+    /**
+     * Formats given linear regression model like "1.0000*x1 + 2.0000*x2 + 3.0000*x3 + 4.0000".
+     *
+     * @param model linear regression model
+     * @return formatted string representation
+     */
+    private static String formatLinearRegressionModelPrettyPrint(LinearRegressionModel model) {
```
Comment (zaleslaw) [inline examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: Better to add into LinearRegressionModel. I'd like the pretty print. Also you could add constraint in 30 or 50 vars to print out this model.

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                Trainer<LinearRegressionModel, Matrix> trainer = new LinearRegressionQRTrainer();
+
+                System.out.println(">>> Perform the training to get the model.");
+                LinearRegressionModel model = trainer.train(distributedMatrix);
+                System.out.println(">>> Linear regression model: " + formatLinearRegressionModelPrettyPrint(model));
+
+                System.out.println(">>> ---------------------------------");
+                System.out.println(">>> | Prediction\t| Ground Truth\t|");
+                System.out.println(">>> ---------------------------------");
+                for (double[] observation : data) {
```
Comment (zaleslaw) [inline examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: The best way here to use LabeledDatasets (but it can be a separate ticket)

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    /**
+     * Sets max number of gradient descent iterations.
+     *
+     * @param maxIterations Max number of gradient descent iterations
+     * @return This gradient descent instance
+     */
+    public GradientDescent withMaxIterations(int maxIterations) {
+        if (maxIterations < 0)
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: Mostly we use asserts to check input arguments, Let's discuss it with @ybabak @YuriBabak

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+     * Sets convergence tolerance.
+     *
+     * @param convergenceTol Condition which decides iteration termination
+     * @return This gradient descent instance
+     */
+    public GradientDescent withConvergenceTol(double convergenceTol) {
+        if (convergenceTol < 0)
+            throw new IllegalArgumentException("Convergence tolerance must be non-negative but got " + convergenceTol);
+        this.convergenceTol = convergenceTol;
+        return this;
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: Mostly we don't use chain and use only setters. Let's discuss it with @ybabak @YuriBabak 
I'd like chain

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    /**
+     * Computes point where loss function takes minimal value.
+     *
+     * @param data Inputs parameters of loss function
+     * @param initialWeights Initial weights
+     * @return Point where loss function takes minimal value
+     */
+    public Vector optimize(Matrix data, Vector initialWeights) {
+        Vector weights = initialWeights, oldWeights = null, oldGradient = null;
+        if (data instanceof SparseDistributedMatrix) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: Could we refactor it to one method to avoid copy-paste?

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    private Vector calculateDistributedGradient(SparseDistributedMatrix data, Vector weights) {
+        SparseDistributedMatrixMapReducer mapReducer = new SparseDistributedMatrixMapReducer(data);
+        return mapReducer.mapReduce(
+            (matrix, args) -> {
+                Matrix inputs = extractInputs(matrix);
+                Vector groundTruth = extractGroundTruth(matrix);
+                return lossGradient.compute(inputs, groundTruth, args);
+            },
+            gradients -> {
+                Vector resultGradient = new DenseLocalOnHeapVector(data.columnSize());
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: What if it is crashed on Local Heap?

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.io.Serializable;
+import java.util.Objects;
+import org.apache.ignite.ml.Exportable;
+import org.apache.ignite.ml.Exporter;
+import org.apache.ignite.ml.Model;
+import org.apache.ignite.ml.math.Vector;
+
+/**
+ * Simple linear regression model which predicts result Y as a linear combination of input variables: Y = b * X.
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java]: Ax+b?

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+ */
+public class LinearRegressionModel implements Model<Vector, Double>, Exportable<LinearRegressionModel>, Serializable {
+    /** */
+    private static final long serialVersionUID = -105984600091550226L;
+
+    /** */
+    private final Vector weights;
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java]: What is this? Please add a good comment

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionQRTrainer.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public class LinearRegressionQRTrainer implements Trainer<LinearRegressionModel, Matrix> {
+    /**
+     * {@inheritDoc}
+     */
+    @Override public LinearRegressionModel train(Matrix data) {
+        long ts1 = System.currentTimeMillis();
+        Vector groundTruth = extractGroundTruth(data);
+        Matrix inputs = extractInputs(data);
+        long ts2 = System.currentTimeMillis();
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionQRTrainer.java]: Don't forget remove these timestamps

File: modules/ml/src/test/java/org/apache/ignite/ml/regressions/linear/ArtificialRegressionDatasets.java [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+            1.57683389508, 1.81535629587, 1.07873616646, -0.355800782882, 2.18333193195, 0.0711071144615},
+        {-165.835194521, -2.76613178307, 0.805314338858, 0.81526046683, -0.710489036197,
+            -1.20189542317, -0.692110074722, -0.117239516622, 1.0431459458, -0.111898596299, -0.0775811519297},
+        {-341.189958588, 0.668555635008, -1.0940034941, -0.497881262778, -0.603682823779,
+            -0.396875163796, -0.849144848521, 0.403936807183, -1.82076277475, -0.137500972546, -1.22769896568}
+    }, new double[] {45.8685095528, 11.9400336005, 16.3984976652, 79.9069814034, 5.65486853464,
+        83.6427296424, 27.4571268153, 73.5881193584, 27.1465364511, 79.4095449062}, -5.14077007134);
+
+    /** */
+    public static class Dataset {
```
Comment (zaleslaw) [inline modules/ml/src/test/java/org/apache/ignite/ml/regressions/linear/ArtificialRegressionDatasets.java]: Please rename this class to TestDataset or something else. Or use LabeledDataset after update

File: modules/yardstick/src/main/ml/org/apache/ignite/yardstick/ml/regression/IgniteOLSMultipleLinearRegressionBenchmark.java [hunk 11]
```diff
@@ -41,52 +41,29 @@
...
-
-        // Check expected beta values from R
-        mdl.estimateRegressionParameters();
-
-        // Check standard errors from R
-        mdl.estimateRegressionParametersStandardErrors();
-
-        // Check expected residuals from R
-        mdl.estimateResiduals();
+        LinearRegressionSGDTrainer trainer = new LinearRegressionSGDTrainer(100_000, 1e-12);
```
Comment (zaleslaw) [inline modules/yardstick/src/main/ml/org/apache/ignite/yardstick/ml/regression/IgniteOLSMultipleLinearRegressionBenchmark.java]: But this was about OLS Regression and you use SGDTrainer. It's strange

File: modules/ml/src/test/resources/datasets/regression/boston.csv [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+23.7, 5.70818, 0.0, 18.1, 0.0, 0.532, 6.75, 74.9, 3.3317, 24.0, 666.0, 20.2, 393.07, 7.74
+25.0, 5.73116, 0.0, 18.1, 0.0, 0.532, 7.061, 77.0, 3.4106, 24.0, 666.0, 20.2, 395.28, 7.01
+21.8, 2.81838, 0.0, 18.1, 0.0, 0.532, 5.762, 40.3, 4.0983, 24.0, 666.0, 20.2, 392.92, 10.42
+20.6, 2.37857, 0.0, 18.1, 0.0, 0.583, 5.871, 41.9, 3.724, 24.0, 666.0, 20.2, 370.73, 13.34
+21.2, 3.67367, 0.0, 18.1, 0.0, 0.583, 6.312, 51.9, 3.9917, 24.0, 666.0, 20.2, 388.62, 10.58
+19.1, 5.69175, 0.0, 18.1, 0.0, 0.583, 6.114, 79.8, 3.5459, 24.0, 666.0, 20.2, 392.68, 14.98
+20.6, 4.83567, 0.0, 18.1, 0.0, 0.583, 5.905, 53.2, 3.1523, 24.0, 666.0, 20.2, 388.22, 11.45
+15.2, 0.15086, 0.0, 27.74, 0.0, 0.609, 5.454, 92.7, 1.8209, 4.0, 711.0, 20.1, 395.09, 18.06
+7.0, 0.18337, 0.0, 27.74, 0.0, 0.609, 5.414, 98.3, 1.7554, 4.0, 711.0, 20.1, 344.05, 23.97
+8.1, 0.20746, 0.0, 27.74, 0.0, 0.609, 5.093, 98.0, 1.8226, 4.0, 711.0, 20.1, 318.43, 29.68
```
Comment (zaleslaw) [inline modules/ml/src/test/resources/datasets/regression/boston.csv]: Don't forget add readme file with short description about attributes, licenses or link to the dataset repository

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            // because we create ignite cache internally.
+            IgniteThread igniteThread = new IgniteThread(ignite.configuration().getIgniteInstanceName(),
+                SparseDistributedMatrixExample.class.getSimpleName(), () -> {
+
+                // Create SparseDistributedMatrix, new cache will be created automagically.
+                System.out.println(">>> Create new SparseDistributedMatrix inside IgniteThread.");
+                SparseDistributedMatrix distributedMatrix = new SparseDistributedMatrix(data);
+
+                System.out.println(">>> Create new linear regression trainer object.");
+                Trainer<LinearRegressionModel, Matrix> trainer = new LinearRegressionQRTrainer();
```
Comment (zaleslaw) [inline examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: OLS could be solved not by QR decomposition only.

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java (hunk 6 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: It's just a resulting vector, it doesn't contain more data than we need. From the other hand it has cardinality equals to cardinality of the feature vector, so it shouldn't be big.

File: modules/ml/src/test/resources/datasets/regression/boston.csv (hunk 12 above)
Comment (dmitrievanthony) [reply modules/ml/src/test/resources/datasets/regression/boston.csv]: README.txt with dataset descriptions added.

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java (hunk 7 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java]: Updated, now it's `Y = weights * X + intercept`. Should be more understandable.

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java (hunk 8 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionModel.java]: Added.

File: modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionQRTrainer.java (hunk 9 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/regressions/linear/LinearRegressionQRTrainer.java]: Removed.

File: modules/ml/src/test/java/org/apache/ignite/ml/regressions/linear/ArtificialRegressionDatasets.java (hunk 10 above)
Comment (dmitrievanthony) [reply modules/ml/src/test/java/org/apache/ignite/ml/regressions/linear/ArtificialRegressionDatasets.java]: Renamed.

File: modules/yardstick/src/main/ml/org/apache/ignite/yardstick/ml/regression/IgniteOLSMultipleLinearRegressionBenchmark.java (hunk 11 above)
Comment (dmitrievanthony) [reply modules/yardstick/src/main/ml/org/apache/ignite/yardstick/ml/regression/IgniteOLSMultipleLinearRegressionBenchmark.java]: Generally speaking `SGDTrainer` also makes `OLS Regression` just because it uses Least Square as a loss function.

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java (hunk 5 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: Good idea, done.

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java (hunk 4 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: As discussed will keep this approach with chain methods.

File: modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java (hunk 3 above)
Comment (dmitrievanthony) [reply modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java]: Replaced with asserts.

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java (hunk 1 above)
Comment (dmitrievanthony) [reply examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: Added into `toString` of `LinearRegressionModel`.

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java (hunk 2 above)
Comment (dmitrievanthony) [reply examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: Good idea, will be done in separate task.

File: examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java (hunk 13 above)
Comment (dmitrievanthony) [reply examples/src/main/ml/org/apache/ignite/examples/ml/regression/linear/DistributedLinearRegressionExample.java]: Added separate examples for `QRTrainer` and `SGDTrainer`.
